### PR TITLE
External configuration via environment variables.

### PR DESCRIPTION
**Pull request by humb1t**
Signed-off-by: Nikita Puzankov <humb1t@yandex.ru>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,8 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+dotenv = "0.15.0"
```
Comment (takemiyamakoto) [inline Cargo.toml]: let's not add more libraries just for reading a simple config file

File: .env [hunk 2]
```diff
@@ -0,0 +1,2 @@
+TORII_URL=127.0.0.1:1337
```
Comment (takemiyamakoto) [inline .env]: I would like to have config as a JSON file called config.json. There will be another JSON file called trusted_peers.json as well (which will have the public keys and IPs of trust peers to bootstrap syncing with the network).

File: src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,6 @@
+#[macro_use]
+extern crate dotenv_codegen;
```
Comment (takemiyamakoto) [inline src/lib.rs]: Not needed since we won't use dotenv

File: src/main.rs [hunk 4]
```diff
@@ -1,22 +1,23 @@
+#[macro_use]
+extern crate dotenv_codegen;
```
Comment (takemiyamakoto) [inline src/main.rs]: Not needed since we won't use dotenv

File: src/main.rs [hunk 5]
```diff
@@ -1,22 +1,23 @@
...
 use std::thread;
 use std::time::Duration;
 
 fn main() {
     println!("Hyperledgerいろは2にようこそ！");
-
     // Setup data structures
-    let block_time_ms = 1000; //TODO: read from config file
-
+    let block_time_ms = dotenv!("BLOCK_BUILD_STEP_MS")
```
Comment (takemiyamakoto) [inline src/main.rs]: Just read in and parse the config.json and trusted_peers.json here at the beginning of main. Then these variables can be used to initialize the iroha node.

File: src/networking/torii.rs [hunk 6]
```diff
@@ -119,8 +120,8 @@ mod tests {
             torii.start();
         });
         std::thread::sleep(std::time::Duration::from_millis(50));
-        let mut stream =
-            std::net::TcpStream::connect(TORII_URL).expect("Failet connect to the server.");
+        let mut stream = std::net::TcpStream::connect(dotenv!("TORII_URL"))
```
Comment (takemiyamakoto) [inline src/networking/torii.rs]: This should be read in from main and set upon initialization of torii.

File: Cargo.toml (hunk 1 above)
Comment (humb1t) [reply Cargo.toml]: ok

File: .env (hunk 2 above)
Comment (humb1t) [reply .env]: As you wish)

File: src/config.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
+use std::{collections::HashMap, env, fs, path::Path};
+
+pub struct Configuration {
+    pub torii_url: String,
+    pub block_build_step_ms: u64,
```
Comment (kamilsa) [inline src/config.rs]: We will add special kind of transaction to set block_build_step. For now I think we can remove it from configs and just hardcode with some value (i.e. 3 secs)

File: benches/torii.rs [hunk 8]
```diff
@@ -29,10 +34,11 @@ fn query_requests(criterion: &mut Criterion) {
     let mut query_request = QUERY_REQUEST_HEADER.to_vec();
     query_request.append(&mut query);
     group.throughput(Throughput::Bytes(query_request.len() as u64));
+    let torii_url = config.torii_url.to_string();
     group.bench_function("query", |b| {
         b.iter(|| {
             let mut stream =
-                std::net::TcpStream::connect(TORII_URL).expect("Failet connect to the server.");
+                std::net::TcpStream::connect(&torii_url).expect("Failet connect to the server.");
```
Comment (takemiyamakoto) [inline benches/torii.rs]: Failed?

File: test.env [hunk 9]
```diff
@@ -0,0 +1,2 @@
+TORII_URL=127.0.0.1:1337
```
Comment (kamilsa) [inline test.env]: I don't think it is a good idea to keep test config in the root, if it is only needed for the test you can move it to the same folder with that test

File: src/config.rs (hunk 7 above)
Comment (humb1t) [reply src/config.rs]: let it stay here and we will replace it with a ISI later?

Comment (kamilsa) [reply src/config.rs]: then leave a todo for that

Comment (humb1t) [reply src/config.rs]: https://soramitsu.atlassian.net/browse/RH2-35

File: test.env (hunk 9 above)
Comment (humb1t) [reply test.env]: Moved into `./tests`.

**Pull request merged**